crates/emitter/src/array_emitter.rs

```diff
@@ -0,0 +1,175 @@
+
+/// Struct for emitting bytecode for an array element.
+pub struct ArrayElementEmitter<'a, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub state: &'a mut ArrayEmitterState,
+    pub elem: F,
+}
+
+impl<'a, F> ArrayElementEmitter<'a, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack] ARRAY INDEX?
+
+        (self.elem)(emitter)?;
+        //              [stack] ARRAY INDEX? ELEM
+
+        match &mut self.state.0 {
+            ArrayEmitterStateInternal::BeforeSpread { ref mut index } => {
+                emitter.emit.init_elem_array(*index);
+                //      [stack] ARRAY
+
+                *index += 1;
+            }
+            ArrayEmitterStateInternal::AfterSpread => {
+                emitter.emit.init_elem_inc();
+                //      [stack] ARRAY INDEX+1
+            }
+        }
+
+        //              [stack] ARRAY INDEX?
+
+        Ok(())
+    }
+}
+
+/// Struct for emitting bytecode for an array element with hole.
+pub struct ArrayElisionEmitter<'a> {
+    pub state: &'a mut ArrayEmitterState,
+}
+
+impl<'a> ArrayElisionEmitter<'a> {
+    pub fn emit(self, emitter: &mut AstEmitter) {
+        //              [stack] ARRAY INDEX?
+
+        emitter.emit.hole();
+        //              [stack] ARRAY INDEX? HOLE
+
+        match &mut self.state.0 {
+            ArrayEmitterStateInternal::BeforeSpread { ref mut index } => {
+                emitter.emit.init_elem_array(*index);
+                //      [stack] ARRAY
+
+                *index += 1;
+            }
+            ArrayEmitterStateInternal::AfterSpread => {
+                emitter.emit.init_elem_inc();
+                //      [stack] ARRAY INDEX+1
+            }
+        }
+
+        //              [stack] ARRAY INDEX?
+    }
+}
+
+/// Struct for emitting bytecode for an array element with spread.
+pub struct ArraySpreadEmitter<'a, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub state: &'a mut ArrayEmitterState,
+    pub elem: F,
+}
+
+impl<'a, F> ArraySpreadEmitter<'a, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack] ARRAY INDEX?
+
+        match self.state.0 {
+            ArrayEmitterStateInternal::BeforeSpread { index } => {
+                emitter.emit.numeric(index as f64);
+            }
+            _ => {}
+        }
+        self.state.0 = ArrayEmitterStateInternal::AfterSpread;
+
+        //              [stack] ARRAY INDEX
+
+        Err(EmitError::NotImplemented("TODO: spread element"))
+    }
+}
+
+enum ArrayEmitterStateInternal {
+    BeforeSpread { index: u32 },
+    AfterSpread,
+}
+
+/// Opaque struct that can be created only by ArrayEmitter.
+/// This guarantees that *PropertyEmitter structs cannot be used outside
```

Comment:
is this comment right? It looks like propertyEmitters are only used with the ObjectEmitter and not the ArrayEmitter? Is this more related to spread?

Reply:
sorry, it should be Array*Emitter